L1Trigger/TrackFindingTMTT/interface/KalmanState.h

```diff
@@ -0,0 +1,118 @@
+                unsigned n_skipped,
+                unsigned kLayer_next,
+                unsigned layerId,
+                const KalmanState *last_state,
+                const std::vector<double> &x,
+                const TMatrixD &pxx,
+                const TMatrixD &K,
+                const TMatrixD &dcov,
+                const StubCluster *stubcl,
+                double chi2rphi,
+                double chi2rz,
+                L1KalmanComb *fitter,
+                GET_TRACK_PARAMS f);
+    KalmanState(const KalmanState &p);
+    ~KalmanState() {}
+
+    KalmanState &operator=(const KalmanState &other);
+
+    unsigned nextLayer() const { return kLayerNext_; }
+    unsigned layerId() const { return layerId_; }
+    unsigned endcapRing() const { return endcapRing_; }
+    bool barrel() const { return barrel_; }
+    unsigned nSkippedLayers() const { return n_skipped_; }
+    // Hit coordinates.
+    double r() const { return r_; }
+    double z() const { return z_; }
+    const KalmanState *last_state() const { return last_state_; }
+    // Helix parameters (1/2R, phi relative to sector, z0, tanLambda)
+    std::vector<double> xa() const { return xa_; }
+    // Covariance matrix on helix params.
+    TMatrixD pxxa() const { return pxxa_; }
+    // Kalman Gain matrix
+    TMatrixD K() const { return K_; }
+    // Hit position covariance matrix.
+    TMatrixD dcov() const { return dcov_; }
+    // Hit
+    const StubCluster *stubCluster() const { return stubCluster_; }
+    double chi2() const { return chi2rphi_ + chi2rz_; }
+    double chi2scaled() const { return chi2rphi_ / kalmanChi2RphiScale_ + chi2rz_; }  // Improves electron performance.
+    double chi2rphi() const { return chi2rphi_; }
+    double chi2rz() const { return chi2rz_; }
+    unsigned nStubLayers() const { return n_stubs_; }
+    L1track3D candidate() const { return l1track3D_; }
+    unsigned int hitPattern() const { return hitPattern_; }  // Bit-encoded KF layers the fitted track has stubs in.
+
+    bool good(const TP *tp) const;
+    double reducedChi2() const;
+    const KalmanState *last_update_state() const;
+    std::vector<const Stub *> stubs() const;
+    L1KalmanComb *fitter() const { return fitter_; }
+    GET_TRACK_PARAMS fXtoTrackParams() const { return fXtoTrackParams_; };
+
+    static bool orderChi2(const KalmanState *left, const KalmanState *right);
+    static bool orderMinSkipChi2(const KalmanState *left, const KalmanState *right);
+
+    static bool order(const KalmanState *left, const KalmanState *right);
+    void dump(ostream &os, const TP *tp = 0, bool all = 0) const;
+    void setChi2(double chi2rphi, double chi2rz) {
+      chi2rphi_ = chi2rphi;
+      chi2rz_ = chi2rz;
+    }
+
+    // If using HLS, note/get additional output produced by HLS core.
+    //void setHLSselect(unsigned int mBinHelix, unsigned int cBinHelix, bool consistent) { mBinHelixHLS_ = mBinHelix; cBinHelixHLS_ = cBinHelix; consistentHLS_ = consistent;}
```

Comment:
An HLS version of the Kalman Filter fit also exists, which provides a fully accurate emulation of the FPGA firmware. To simplify the CMSSW 11.1 release, this has been removed from the repository. But it is likely to be added to future releases (as an external library), so it would be useful to keep the calls to HLS code this present in the L1 track C++ code (albeit commented out), so we can reinstate them easily in future.